File: Contest/SolvingQuestionsWithBrainpower.java
class Solution {
	// TC : O(n)
	// SC:  O(n)
	public long mostPoints(int[][] questions) {
		long[] dp = new long[questions.length];

		for(int i=questions.length-1;i>=0;i--){
			int furtureReachIndex = i + questions[i][1] +1;
			if(furtureReachIndex<questions.length){
				dp[i] = dp[furtureReachIndex] + questions[i][0];
			} else {
				dp[i] = questions[i][0];
			}

			if(i<questions.length-1) {
				// skip the ith index
				dp[i] = Math.max(dp[i+1], dp[i]);
			}
		}
		return dp[0];

	}
}

File: January2022/C++/Solving_Questions_With_Brainpower.cpp
class Solution {
public:
    long long mostPoints(vector<vector<int>>& arr) {
        long long n = arr.size();
        vector<long long> dp(n,0);
        for(long i = n-1;i>=0;i--){
                dp[i]=arr[i][0];
            if((arr[i][1]+i+1)<n)
                dp[i]=max(dp[i],dp[arr[i][1]+i+1]+(arr[i][0]));
            if(i+1<n)
                dp[i]=max(dp[i],dp[i+1]);
        }
        return dp[0];
    }
};
